### PR TITLE
Created command context provider to set job group ID as part of run tags

### PR DESCRIPTION
**Pull request by jerrylian-db**
## What changes are proposed in this pull request?

Created command context provider to set job group ID as part of run tags. We use the job group ID as the command ID because it is set for every command in Databricks.

## How is this patch tested?

Unit tests were added for the new run context.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/utils/mlflow_tags.py [hunk 1]
```diff
@@ -27,6 +27,7 @@
 MLFLOW_DATABRICKS_WEBAPP_URL = "mlflow.databricks.webappURL"
 MLFLOW_DATABRICKS_RUN_URL = "mlflow.databricks.runURL"
 MLFLOW_DATABRICKS_CLUSTER_ID = "mlflow.databricks.cluster.id"
+MLFLOW_DATABRICKS_COMMAND_ID = "mlflow.databricks.commandID"
```
Comment (jerrylian-db) [inline mlflow/utils/mlflow_tags.py:30]: Not sure whether to rename this to `"mlflow.databricks.jobGroup.id"`.

File: mlflow/utils/databricks_utils.py [hunk 2]
```diff
@@ -151,6 +151,16 @@ def get_cluster_id():
     return spark_session.conf.get("spark.databricks.clusterUsageTags.clusterId")
 
 
+def get_job_group_id():
+    try:
+        dbutils = _get_dbutils()
+        job_group_id = dbutils.entry_point.getJobGroupId()
```
Comment (jerrylian-db) [inline mlflow/utils/databricks_utils.py:157]: I found this function [here](https://github.com/databricks/universe/blob/9676a1c16f1c3fdb10da3f19fe01df907602ef42/spark/driver/python/python_shell/dbruntime/dbutils.py#L141) and decided to use it. Let me know if more options are needed.

Comment (jerrylian-db) [inline mlflow/utils/databricks_utils.py:157]: I found this method in the dbutils.py file. Let me know if this is sufficient to cover all environments.

File: mlflow/utils/mlflow_tags.py (hunk 1 above)
Comment (smurching) [reply mlflow/utils/mlflow_tags.py:30]: The current naming seems fine (basically semantically we want this ID to be the unique ID of a command execution in a Databricks notebook). Maybe `mlflow.databricks.notebook.commandId` for clarity? And we could add a clarifying comment to ^that effect explaining what this tag is

File: mlflow/utils/databricks_utils.py (hunk 2 above)
Comment (smurching) [reply mlflow/utils/databricks_utils.py:157]: From offline discussion, this seems like the right method to use / should be allowlisted in all cluster environments! Ideally we can cover this our tests too

**Review by smurching — COMMENTED** [review]
LGTM pending manual verification that an appropriate tag is set when running the MLflow python client built from this PR against DBR :)

**Comment by jerrylian-db** [comment]
I've manually tested the change on a high concurrency cluster in Databricks. Below is the screenshot:
Screen Shot 2021-07-01 at 4 44 36 PM

File: tests/tracking/context/test_databricks_command_context.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+def test_databricks_command_run_context_tags():
+    with mock.patch("mlflow.utils.databricks_utils.get_job_group_id") as job_group_id_mock:
+        assert DatabricksCommandRunContext().tags() == {
+            MLFLOW_DATABRICKS_NOTEBOOK_COMMAND_ID: job_group_id_mock.return_value
+        }
+
+
+def test_databricks_command_run_context_tags_nones():
+    with mock.patch("mlflow.utils.databricks_utils.get_job_group_id", return_value=None) as _:
```
Comment (harupy) [inline tests/tracking/context/test_databricks_command_context.py:22]:
```suggestion
    with mock.patch("mlflow.utils.databricks_utils.get_job_group_id", return_value=None):
```

nit: we can remove `as _`

File: tests/tracking/context/test_databricks_command_context.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+from mlflow.utils.mlflow_tags import MLFLOW_DATABRICKS_NOTEBOOK_COMMAND_ID
+from mlflow.tracking.context.databricks_command_context import DatabricksCommandRunContext
+
+
+def test_databricks_command_run_context_in_context():
+    with mock.patch("mlflow.utils.databricks_utils.get_job_group_id") as job_group_id_mock:
+        assert DatabricksCommandRunContext().in_context() == (
+            job_group_id_mock.return_value is not None
+        )
```
Comment (harupy) [inline tests/tracking/context/test_databricks_command_context.py:11]:
```suggestion
    with mock.patch("mlflow.utils.databricks_utils.get_job_group_id", return_value="1") as job_group_id_mock:
        assert DatabricksCommandRunContext().in_context()
```

Can we explicitly specify `return_value` here?

**Comment by harupy** [comment]
The `build_doc_r` failure should be fixed by merging the master branch.

**Review by smurching — APPROVED** [review]
LGTM to merge after @harupy 's comments are addressed!

**Pull request merged**